HTML <script> type=importmap prep

hamishwillee opened the pull request:
This does some initial prep work for documenting "importmaps". Specifically it updates the first mention of importmap so that I can discuss next steps in a proper PR.
It also fixes some layout and highlights some sections that might need work.

This is part of #16150

- [ ] When this goes in, merge #22888

hamishwillee commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -147,13 +132,23 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
+      <!-- For information on `importmap` see the guide: [Importing JavaScript Modules into HTML](/en-US/docs/Web/HTML/Importing_JavaScript_Modules). -->
+    - **Any other value:** The embedded content is treated as a data block, and won't be processed by the browser.
+      Developers must use a valid MIME type that is not a JavaScript MIME type to denote data blocks.
+      All of the other attributes will be ignored, including the `src` attribute.
```
@Josh-Cena The "including the `src` attribute." confuses me a bit. If it is true that the `src` attribute is ignored, how can a module import JSON code from a file?

Josh-Cena replied:
@hamishwillee It doesn't make sense to import JSON through `<script>`. At the end of the day, linking a module into your HTML must have side-effects, because otherwise its data is not visible to anything else. However, JSON is entirely side-effect-free, so putting it inside `<script>` does nothing.

The idea of an import map is to provide _specifier resolutions_. For example, when you have this within a JS module:

```js
import * as Three from "three";
```

How does the browser know where to find `three`? It could treat that as a relative path, but that's not how Node does things. So import maps provide this custom resolution layer. (See https://threejs.org/docs/index.html#manual/en/introduction/Installation, where you resolve `three` to a CDN URL)

hamishwillee replied:
Thanks @Josh-Cena 
I'm now off until Friday, so I will think on this and get back to you. 
I think do understand what this is for - I've read both the explainer and the spec now (they differ mostly in that the spec only allows import maps "inline"). Mostly this was my way of engaging _you_ in the discussion! Appreciate the summary though - I'll probably steal it :-)

Half of me thinks this should go in JavaScript docs because it is of most interest to JavaScript developers coming from node (or somewhere else that has dynamic resolution of the specifier). The purist in me says that this should be in HTML because it is browser specific and shouldn't be in any great detail in "JavaScript Modules". As you say, it depends on the amount of content. I think I might try write it in HTML and we might then agree it needs to move. When you have the content that is easy enough.

hamishwillee replied:
@Josh-Cena OK, I've thought about it some more. I think it is worthy of a topic, probably in the JavaScript section and linked from HTML script docs. At the moment my working title is "Import maps: Javascript module resolution in Browsers".

I was wondering if you could confirm my understanding of one aspect. In node you specify your dependencies in a package.json and these are fetched locally. You can then specify a bare import path, and node knows this means "relative to a particular file system location" and can resolve the module.

Normally if you wanted to run this code in a browser you might build the code to a single JS file and import this into your script. You would have to do this because if you don't build the whole thing your browser wouldn't be able to resolve the dependencies of your module that are specified as bare paths. Right?

But now you can use importmap. In this case you would have to specify an alias for each of the modules that you import as bare modules, and of any modules that they also import. Is that right? Probably that is fairly easy because you could still host these all in a common tree on your server so you might only need one alias.

Is that the main use case of this. Or is it more "node developers are going to write code that uses bare imports, so lets allow them to do this" - because we're nice?

hamishwillee replied:
PS I guess I am after a sentence "you will use this feature when you want to Xyz". Which might be "you will use this feature when you want to import or implement modules that can easily be reused in both server and browser-side".

Further, I'm wondering if this then implies for the person developing the module that they need to publish an appropriate import map structure.

Josh-Cena replied:
> You would have to do this because if you don't build the whole thing your browser wouldn't be able to resolve the dependencies of your module that are specified as bare paths.

Yep, this is right, or at least it would be hard to set up the correct serving paths.

> Is that (making dual use node/browser modules) the main use case of this? Or is it more "node developers are going to write code that uses bare imports, so lets allow them to do this" - because we're nice?

Frankly I don't know... I'm not familiar with standardization processes outside the ES spec so I don't know where to look for motivations, but I'd say that using bare specifiers in general is much more ergonomic than using fully qualified URLs everywhere. Maybe it helps with Node/browser interoperability as well.

hamishwillee replied:
Thanks, appreciate the feedback. I tend to like to start with reasons, but I can be creative.

hamishwillee commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -147,13 +132,23 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
       Unlike classic scripts, module scripts require the use of the CORS protocol for cross-origin fetching.
-    - **Any other value:** The embedded content is treated as a data block which won't be processed by the browser. Developers must use a valid MIME type that is not a JavaScript MIME type to denote data blocks. The `src` attribute will be ignored.
+    - **`importmap`:** This value indicates that the body of the element defines an import map, which allows greater control over which URLS are fetched when importing modules.
+      <!-- For information on `importmap` see the guide: [Importing JavaScript Modules into HTML](/en-US/docs/Web/HTML/Importing_JavaScript_Modules). -->
```
Is it reasonable to create the document "Importing JavaScript Modules into HTML" here as I am suggesting - i.e. a new guide topic? Is it the "right" guide topic - ie should it be "Importing Javascript into HTML" and cover all kinds of importing?

Josh-Cena replied:
This is the fun part—modules are where HTML meets JavaScript, so I'm not sure if we want to put this as a section of the JS modules guide, or create a new guide under HTML (where we can also document plain `<script>` tags, I suppose?) It mainly depends on how much stuff there is to talk about.

hamishwillee commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -226,11 +221,13 @@ You can also use the `<script>` element to embed data in HTML with server-side r
 
-In older browsers that don't support the `async` attribute, parser-inserted scripts block the parser; script-inserted scripts execute asynchronously in IE and WebKit, but synchronously in Opera and pre-4 Firefox. In Firefox 4, the `async` DOM property defaults to `true` for script-created scripts, so the default behavior matches the behavior of IE and WebKit.
+In older browsers that don't support the `async` attribute, parser-inserted scripts block the parser; script-inserted scripts execute asynchronously in IE and WebKit, but synchronously in Opera and pre-4 Firefox.
+In Firefox 4, the `async` DOM property defaults to `true` for script-created scripts, so the default behavior matches the behavior of IE and WebKit.
 
 To request script-inserted external scripts be executed in the insertion order in browsers where the `document.createElement("script").async` evaluates to `true` (such as Firefox 4), set `async="false"` on the scripts you want to maintain order.
 
-> **Warning:** Never call `document.write()` from an async script. In Firefox 3.6, calling `document.write()` has an unpredictable effect. In Firefox 4, calling `document.write()` from an async script has no effect (other than printing a warning to the error console).
+> **Warning:** Never call `document.write()` from an async script.
+> In Firefox 3.6, calling `document.write()` has an unpredictable effect. In Firefox 4, calling `document.write()` from an async script has no effect (other than printing a warning to the error console).
```
The text  seems to refer to  scripts inserted by javascript code. But this doc is about scripts inserted using the `<script>` element. Secondly, text about "document.write()" should presumably belong in that document. 

Is any of this relevant here or could it just be deleted? If not, what could move into BCD?

Rumyra replied:
This seems superfluous to me _and_ outdated - I think delete the section and add anything relevant to bcd

hamishwillee replied:
I agreed. Removed.

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -147,13 +132,23 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
       For information on using `module`, see our [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules) guide.
       Unlike classic scripts, module scripts require the use of the CORS protocol for cross-origin fetching.
-    - **Any other value:** The embedded content is treated as a data block which won't be processed by the browser. Developers must use a valid MIME type that is not a JavaScript MIME type to denote data blocks. The `src` attribute will be ignored.
+    - **`importmap`:** This value indicates that the body of the element defines an import map, which allows greater control over which URLS are fetched when importing modules.
```
```suggestion
    - **`importmap`:** This value indicates that the body of the element defines an import map, which allows greater control over which URLs are fetched when importing modules.
```

hamishwillee commented on files/en-us/web/html/script_import_maps/index.md:
```diff
@@ -0,0 +1,69 @@
+---
```
Note, added this and reverted to draft. This can no longer be merged until I have finished. This is WIP - intro give me a bit of writer's block.

hamishwillee replied:
@Josh-Cena @wbamberg I was wondering if you two could have a scan of this PR for "structural common sense".
It is not done, but I've been having some trouble finding a structure that works, and I don't want to progress too far if you hate this.

Anyway, what I have ended is doing a reference-like guide. I could make it more officially reference by putting it under the element/script tree. The main reason I did it this way is that 
- it feels quite different to all other elements.
- it would in theory have to be deep in a nested reference structure "i.e. it is script element, type attribute, importmap value. The title of that would not be useful if you wanted to find the topic - the current title is better for search.

I could of course try split into guide and reference. I think that would be repetetive.

Do you think this approach is workable?

hamishwillee commented on files/en-us/web/html/script_import_maps/index.md:
```diff
@@ -0,0 +1,218 @@
+Particularly in larger applications, it is common for code to depend on different versions of the same module.
+Using relative ...
+
+### Feature detection
```
Note, I am not sure how useful feature detection is. Yes this works, but I don't know how you can sensibly fall back to anything with static maps?

Maybe this would be useful if you were using dynamic import map import where you construct the map?

But can

hamishwillee commented on files/en-us/web/html/script_import_maps/index.md:
````diff
@@ -0,0 +1,218 @@
+#### Syntax
+
+```json
+/* Default imports for bare modules and paths */
````
I think these are red because JSON doesn't support comments natively (I don't think). Is that a problem/do we deal with this in a standard way?

Josh-Cena replied:
Typically, we would use `jsonc` as the language tag; unfortunately that's not supported in Yari. In this case, I would stay with `json`, because the render output is fine, just that GitHub is unhappy.

wbamberg commented:
> @Josh-Cena @wbamberg I was wondering if you two could have a scan of this PR for "structural common sense".

I think either of the following would be better:
- incorporate "Import map JSON representation" into "Import maps: Javascript module resolution in Browsers", and live with the fact that we have reference-y material in a guide page
- have a separate page for the HTML `type` attribute under https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes, and document import map syntax in there

Probably the second is better.

It's an IA problem that we don't know what kind of page https://pr22300.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/script/importmap_json_representation is, and exceptions like this will make it much harder for us to drive things like sidebar generation or page structure validation.

Also I don't like guide-y pages having "Specifications" or "Browser compatibility" sections. Mike says it better than me in this comment: https://github.com/mdn/yari/pull/5971#issuecomment-1103975670. One concrete consequence of that is that the compat at https://pr22300.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/script/importmap_json_representation#browser_compatibility is not actually for the import map, it's for the `importmap` attribute itself. If the syntax changes, the compat story for the syntax will be different, but the BCD table will (correctly but misleadingly in that context) stay the same.

hamishwillee commented:
Thanks @wbamberg 

> > @Josh-Cena @wbamberg I was wondering if you two could have a scan of this PR for "structural common sense".
> 
> I think either of the following would be better:
> 
> * incorporate "Import map JSON representation" into "Import maps: Javascript module resolution in Browsers", and live with the fact that we have reference-y material in a guide page
> * have a separate page for the HTML `type` attribute under https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes, and document import map syntax in there
> 
> Probably the second is better.

Respectfully, I think I disagree:
- I just split that out of "Import maps: Javascript module resolution in Browsers" because you correctly pointed out that this is reference masquerading as guide. At least in a separate document it is clearly reference.
- Having it in HTML `type` attribute is possible; I see some benefit because it is a natural extension of our nesting structure. But it is still mixing docs about a JSON format in with HTML attributes. 

The reason I split this into its own reference doc [Element > Script> Importmap json representation](https://pr22300.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/script/importmap_json_representation) are:
- Discoverability - it's visible. If this is pushed into `Element > Script > type` it will be nested pretty deep
- Separation of concerns - the JSON and HTML are quite different.
- Ability to provide clearly separate compatibility info for the format (though BCD required eventually for this, if it ever changes).

Is that convincing? If not, then I will try the HTML `type` approach.

> I don't like guide-y pages having "Specifications" or "Browser compatibility" sections. Mike says it better than me in this comment: [mdn/yari#5971 (comment)](https://github.com/mdn/yari/pull/5971#issuecomment-1103975670). 

Yes, that is why  [Element > Script> Importmap json representation](https://pr22300.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/script/importmap_json_representation) is tagged as _Reference_. It is intended to be reference.

The question is really whether this is separate or part of `script > type`.

I do plan on keeping the compat table in the guide "for now". When support for a feature is not wide, that is useful. I'd take direction in review if you felt strongly it was not useful.

> One concrete consequence of that is that the compat at https://pr22300.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/script/importmap_json_representation#browser_compatibility is not actually for the import map, it's for the `importmap` attribute itself. If the syntax changes, the compat story for the syntax will be different, but the BCD table will (correctly but misleadingly in that context) stay the same.

Correct. For now they are the same and it won't become a problem until there is a compatibility change in the JSON. I actually think that relatively unlikely.

wbamberg commented:
> I have tried to match your suggestion here.

Thanks for this Hamish, and sorry to be slow responding. This is connected to https://github.com/mdn/content/issues/22705 in which I did unearth one existing example of an element/attribute/value page: https://developer.mozilla.org/en-US/docs/web/html/element/meta/name/theme-color . Also the `<input type=*` pages we have are kind of examples of this in disguise.

I think from a language-structural and IA point of view this makes sense. I guess it makes sense that we sometimes need the space to break out attribute values into separate pages, and having them underneath the attributes seems reasonable too. I'm a little unhappy still that there's only one previous example of this, but like I say it makes sense to me.

hamishwillee commented:
Thanks @wbamberg . We'll need to fix the page breadcrumb, but I think it works OK. It is certainly no worse.
Now the structure is broadly agreed I can hopefully just finish this off!

wbamberg commented:
> Thanks @wbamberg . We'll need to fix the page breadcrumb, but I think it works OK. It is certainly no worse. Now the structure is broadly agreed I can hopefully just finish this off!

Oh sorry, I forgot to mention the breadcrumbs!  Yes, I believe there is some dodgy code in Yari that tries to derive "short titles" from breadcrumbs, and it's obviously fragile. Having a `short-title` front matter property would solve this.

I expect if you chose a different title format it would work better.

TBH I don't love the title in these pages. I think just "type" and "importmap" would be OK, or better still "<script>: type attribute" and "<script type="importmap>" (which aligns with our choice for the attribute-names-in-disguise like https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email ).

But how to title these things is  whole other conversation that doesn't IMO have to block this.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/javascript_modules_in_browsers/index.md:
```diff
@@ -0,0 +1,102 @@
+  - HTML
+---
+
+{{HTMLSidebar}}
```
Feels weird to see the HTML sidebar in a JS guide, but modules are indeed in the gray zone...

hamishwillee replied:
@Josh-Cena No, you're right. This will be a JS guide with HTML flavour. So that should change.

It might even be worth merging this into the JS Modules doc, with caveats. But for now I am keeping it clearly separate.

hamishwillee commented:
@Josh-Cena @wbamberg I've marked this ready for review. Would appreciate your comments. I've rushed a bit today to get something that can be looked at. 

This draws on the explainer https://github.com/WICG/import-maps/blob/main/README.md and spec . Some of my examples are thinly veiled copies. I need to move on though, and I hope this is good enough to get people started. I've put a few comments inline which I hope you might be able to help me with.

hamishwillee commented on files/en-us/web/html/element/script/index.md:
````diff
@@ -199,6 +198,39 @@ Browsers that support the `module` value for the {{htmlattrxref("type", "script"
 <script nomodule src="fallback.js"></script>
 ```
 
+### Importing modules with importmap
````
FYI largely duplicated example of using a bare module, with linking to the new browser doc for more.

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,214 @@
+---
+title: '<script type="importmap">'
```
@wbamberg Note, name(s) changed to match your suggestions.

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
````diff
@@ -0,0 +1,214 @@
+}
+```
+
+<!-- 
````
This comes from the spec example here: https://html.spec.whatwg.org/multipage/webappapis.html#example-import-map-url-like-specifier

I couldn't quite get my head around it. I think it means "if you have `import "./foo/../js/app.mjs"` then the modifier gets resolved first - so this becomes `import "./js/app.mjs"` and will therefore match a specifier like `./js/app.mjs`.
But then there was some discussion that it might also match something prefixed with `../`. 
Upshot, I am not sure how necessary it is to say this, or follow it up?

hamishwillee commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +152,36 @@ reportPerimeter(square1.length, reportList);
 
 > **Note:** Although imported features are available in the file, they are read only views of the feature that was exported. You cannot change the variable that was imported, but you can still modify properties similar to `const`. Additionally, these features are imported as live bindings, meaning that they can change in value even if you cannot modify the binding unlike `const`.
 
+### Importing bare names
```
So this is update to modules. This is IMO the main use case of import maps, so here we have the same example again. I think it is worth having this here - it has proper caveats that this requires import maps.

hamishwillee commented on files/en-us/web/javascript/guide/modules/javascript_modules_in_browsers/index.md:
```diff
@@ -0,0 +1,217 @@
+
+### General URL remapping
+
+A module specifier key doesn't have to be path — it can also be an absolute URL (or a URL-like relative path like `./`, `../`, `/`).
```
FYI I haven't gone into detail here - it's something you can do, but it isn't something that the spec or explainer particularly recommends.

hamishwillee commented on files/en-us/web/javascript/guide/modules/javascript_modules_in_browsers/index.md:
```diff
@@ -0,0 +1,217 @@
+
+This works in the same way with respect to import maps and has the same restrictions.
+For example, the import must happen before any other module importing has started, and the module specifier must be resolvable to an absolute URL address for a module.
+
```
This is a case where I ran out of puff. I started from https://github.com/WICG/import-maps/blob/main/README.md#dynamic-import-map-example and thought - creating an example that does the same thing but isn't a copy would be a pain. Interested if you think it is needed or we can do without?

hamishwillee commented on files/en-us/web/javascript/guide/modules/javascript_modules_in_browsers/index.md:
```diff
@@ -0,0 +1,217 @@
+This works in the same way with respect to import maps and has the same restrictions.
+For example, the import must happen before any other module importing has started, and the module specifier must be resolvable to an absolute URL address for a module.
+
+### Feature detection
```
This was in the explainer here: https://github.com/WICG/import-maps/blob/main/README.md#feature-detection

I'm not sure how useful it is - if the import map didn't load then to do something useful we'd have to fall back to relative  imports - and if we've done that work, why use this at all?

wbamberg commented on files/en-us/web/javascript/guide/modules/javascript_modules_in_browsers/index.md:
```diff
@@ -0,0 +1,217 @@
+
+## See also
+
+- [`<script>` HTML element > `type` attribute](/en-US/docs/Web/HTML/Element/script#attr-type)
```
```suggestion
- [The `type` attribute of HTML `<script>` elements](/en-US/docs/Web/HTML/Element/script#attr-type)
```

wbamberg commented on files/en-us/web/javascript/guide/modules/javascript_modules_in_browsers/index.md:
```diff
@@ -0,0 +1,217 @@
+- [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import)
+- [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import)
+
+{{Previous("Web/JavaScript/Guide/Modules")}}
```
I'm not sure this is a blocker but I wonder how this page fits into the JS IA and in particular the JS Guide. We have a `Previous` link here that goes to the [JS module page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules), and that page is [the last page in the JavaScript Guide](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide#javascript_modules). (see also the JS sidebar, which lists all the chapters in the JS guide).

Give the fact that this is a separate page and it has a `Previous` link pointing to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules, one would expect:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules to have a `Next` pointing back to this page
- this page to live at the same level of the hierarchy as https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules and the other chapters of the Guide (with all the things that flow from that, such as it appearing in the sidebar after the "Modules" page).

But it doesn't, and it shouldn't of course because it is really a sub-chapter of the "Modules" page, not a next chapter at all.

But I don't think the Guide is really organized like that, with sub-chapters in separate pages. The guide expects each chapter to be self-contained in a single page (AFAICT). What would be correct, in terms of the organization of the Guide, would be for this page to be a subsection of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules, just like, say, https://pr22300.content.dev.mdn.mozit.cloud/en-US/docs/Web/JavaScript/Guide/Modules#dynamic_module_loading, and referenced in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide#javascript_modules along with the other subsections of the "Modules" chapter.

Josh-Cena replied:
One problem I have with the modules guide is that it's too long—and rightfully so, because it covers so many materials—and it sort of conflates how JS modules work with how JS modules work in browsers. That may or may not be an issue, but Node ESM works quite differently from browsers. As such I would prefer this to be separate, and we can eventually move some of the web-server-related stuff from the modules guide to here as well.

hamishwillee replied:
Thanks. To summarize, our options are:
1. Move content into Modules doc
2. Fix up linking such that this is a peer of modules - i.e. make as last topic in sequence.
3. Make a proper sub topic.

The right thing for me to do **now** is merge this into "JavaScript Modules". I appreciate this is not optimal @Josh-Cena (which is why I wrote it as a sub topic) but it is better than me creating a new approach for JavaScript docs on the fly.

When you come to restructure this then all the information will still be present.

wbamberg reviewed (changes requested):
Thanks Hamish! Lots of nit-picking changes but overall it's really clear. I did have one gnarlier question about how the JS page fits into the JS Guide structure.

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,214 @@
+
+The **`importmap`** value of the [`type`](/en-US/docs/Web/HTML/Element/script/type) attribute of the [`<script>` element](/en-US/docs/Web/HTML/Element/script) indicates that the body of the element contains an import map.
+
+An import map is a JSON object that allows developers to control what URLs are fetch when importing a [JavaScript module](/en-US/docs/Web/JavaScript/Guide/Modules).
```
```suggestion
An import map is a JSON object that allows developers to control how the browser resolves module specifiers when importing [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules).
```

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,214 @@
+When importing a JavaScript module [JavaScript module](/en-US/docs/Web/JavaScript/Guide/Modules), both the [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import) and [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import) have a "module specifier" that indicates the module to be imported.
+A browser must be able to resolve this specifier to an absolute URL in order to be able to import the module.
+
+For example, the following statements in a module `<script>` element import elements from the module specifier "./modules/shapes/square.js", which is a path relative to the base URL of the document, and the module specifier "https://example.com/shapes/circle.js", which is an absolute path to an URL.
```
```suggestion
For example, the following statements import elements from the module specifier "./modules/shapes/square.js", which is a path relative to the base URL of the document, and the module specifier "https://example.com/shapes/circle.js", which is an absolute URL.
```

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,214 @@
+        - If multiple keys can match the module specifier, the most specific key is used (in other words a specifier "olive/branch/" would match before "olive/").
+
+- `scopes` {{optional_inline}}
+  - : Scopes define path-specific [module specifier maps](#module_specifier_map), allowing the map that is used to depend on the path of the code importing the module.
```
Not sure, but I like your phrasing better.

hamishwillee commented on files/en-us/web/html/element/script/type/index.md:
```diff
@@ -0,0 +1,48 @@
+    Unlike classic scripts, module scripts require the use of the CORS protocol for cross-origin fetching.
+- [`importmap`](/en-US/docs/Web/HTML/Element/script/type/importmap)
+  - : This value indicates that the body of the element contains an import map.
+      This is a JSON object that developers can use to control the URLs that are fetched when importing modules.
```
```suggestion
      This is a JSON object that developers can use to control how the browser resolves module specifiers when importing [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules).
```

hamishwillee commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -145,15 +130,28 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
+
+  - : This attribute indicates the type of script represented. The value of this attribute will be in one of the following:
+
+    - **Omitted, empty string, or a JavaScript MIME type**
```
Fair enough - though I like listing all the cases which are "classic script" together. So I think a better solution is to make it obvious that we mean "no attribute

```suggestion
    - **The attribute is not specified (default), an empty string, or a JavaScript MIME type**
```

hamishwillee commented on files/en-us/web/html/element/script/type/index.md:
```diff
@@ -0,0 +1,48 @@
+
+The value of this attribute indicates the type of data represented by the script, and will be in one of the following:
+
+- **Omitted, empty string, or a JavaScript MIME type**
```
```suggestion
- **The attribute is not specified (default), an empty string, or a JavaScript MIME type**
```

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,209 @@
+
+If there are several module specifier keys in the module specifier map that might match, then the most specific key will be selected (i.e. the one with the longer path/value).
+
+Note that a module specifier like `"./foo/../js/app.js` would match with the module specifier key `./js/app.js`, even though the key text is not exactly the same as the module specifier text. This is because the relative path in the module specifier (`"./foo/../js/app.js`) is resolved to `./js/app.js` before comparing with the import map.
```
@wbamberg This is the paragraph that replaces the bits you commented on in https://github.com/mdn/content/pull/22300/files#r1044974265 and https://github.com/mdn/content/pull/22300/files#r1044969105.

It gets rid of the awkward terminology "post canononicalization". 

I am no longer sure what my original concern with the wording was.

hamishwillee committed:
Recreated files - fix broken builds

hamishwillee commented:
This might be the shittest day ever. Just tried to repair broken update which was failing tests and did a force rebase over my content. I am swearing inside.

hamishwillee committed:
Fix up the docs to match WillB feedback:

hamishwillee committed:
Update JavaScript modules

hamishwillee commented:
All done, again. HOpe it meets with your approval. As per https://github.com/mdn/content/pull/22300#discussion_r1045311102 the guide is merged into JavaScript modules topic. Not ideal given the size of the doc, but consistent.

Josh-Cena commented:
Don't worry about the size of that doc for now—next I'll do a pass through it and hopefully both shorten the content and split it into two pages (platform-agnostic and browser-specific)

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -170,7 +168,8 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
 Scripts without {{HTMLAttrxRef("async", "script")}}, {{HTMLAttrxRef("defer", "script")}} or `type="module"` attributes, as well as inline scripts without the `type="module"` attribute, are fetched and executed immediately, before the browser continues to parse the page.
 
-The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`); a video type (`video/*`); an audio (`audio/*`) type; or `text/csv`. If the script is blocked, an {{domxref("Element/error_event", "error")}} is sent to the element, if not a {{domxref("Element/load_event", "load")}} event is sent.
+The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`), a video type (`video/*`), an audio (`audio/*`) type, or `text/csv`.
```
```suggestion
The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`), a video type (`video/*`), an audio type (`audio/*`), or `text/csv`.
```

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -170,7 +168,8 @@ This element includes the [global attributes](/en-US/docs/Web/HTML/Global_attrib
 
-The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`); a video type (`video/*`); an audio (`audio/*`) type; or `text/csv`. If the script is blocked, an {{domxref("Element/error_event", "error")}} is sent to the element, if not a {{domxref("Element/load_event", "load")}} event is sent.
+The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`), a video type (`video/*`), an audio (`audio/*`) type, or `text/csv`.
+If the script is blocked, an {{domxref("Element/error_event", "error")}} is sent to the element, if not a {{domxref("Element/load_event", "load")}} event is sent.
```
```suggestion
If the script is blocked, an {{domxref("Element/error_event", "error")}} event is sent to the element; otherwise, a {{domxref("Element/load_event", "load")}} event is sent.
```

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -199,6 +198,39 @@ Browsers that support the `module` value for the {{htmlattrxref("type", "script"
 
+### Importing modules with importmap
+
+When importing modules in scripts, browsers that don't support the [`importmap`](#importmap) value for the {{htmlattrxref("type", "script")}} attribute must set a module specifier that can be resolved to an absolute URL.
```
```suggestion
When importing modules in scripts, if you don't use the [`type=importmap`](#importmap) feature, then each module must be specified with a specifier that can be resolved to an absolute URL.
```

hamishwillee replied:
I don't like this as much ^^^. What was wrong with the original? In particular "specified with a specifier" not so pleasant.

Josh-Cena replied:
1. "Browsers that don't support X" typically don't age well, and it's better if we can avoid them.
2. This is not really a support issue—it's just comparing what happens if you use the import map feature versus when you don't. Sort of like the motivation of "why you should use it", and that is applicable for learners 10 years later, even when it's well-supported everywhere.

hamishwillee replied:
@Josh-Cena Fair enough - in particular for API ref where you can find support statements easily.

This then is slightly closer to what I want to say. (will merge it)

```suggestion
When importing modules in scripts, if you don't use the [`type=importmap`](#importmap) feature, then each module must be imported using a module specifier that is either an absolute or relative URL.
```

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
```diff
@@ -199,6 +198,39 @@ Browsers that support the `module` value for the {{htmlattrxref("type", "script"
+### Importing modules with importmap
+
+When importing modules in scripts, browsers that don't support the [`importmap`](#importmap) value for the {{htmlattrxref("type", "script")}} attribute must set a module specifier that can be resolved to an absolute URL.
+For example, below the first module specifier ("./shapes/square.js") resolves relative to the base URL of the document, while the second is an absolute URL.
```
```suggestion
In the example below, the first module specifier ("./shapes/square.js") resolves relative to the base URL of the document, while the second is an absolute URL.
```

(Will's suggestion doesn't seem applied)

hamishwillee replied:
Thanks!

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
````diff
@@ -199,6 +198,39 @@ Browsers that support the `module` value for the {{htmlattrxref("type", "script"
+```
+
+An import map allows you to provide a mapping that, if matched, can replace the text in the module specifier.
+In the example below, the import map below defines keys `square` and `circle` that can be used as an alias for the module specifiers shown above.
````
```suggestion
The import map below defines keys `square` and `circle` that can be used as aliases for the module specifiers shown above.
```

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
````diff
@@ -199,6 +198,39 @@ Browsers that support the `module` value for the {{htmlattrxref("type", "script"
+</script>
+```
+
+This allows us to import bare modules in our scripts:
````
```suggestion
This allows us to import modules with names that appear as neither absolute nor relative:
```

hamishwillee replied:
```suggestion
This allows us to import modules with names rather than absolute or relative URLs.
```

Josh-Cena commented on files/en-us/web/html/element/script/index.md:
````diff
@@ -199,6 +198,39 @@ Browsers that support the `module` value for the {{htmlattrxref("type", "script"
+import { name as circleName } from "circle";
+```
+
+For more examples of what you can do with import maps see: [JavaScript modules > Importing modules using import maps](/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps).
````
```suggestion
For more examples of what you can do with import maps, see the [Importing modules using import maps](/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps) section in the JavaScript modules guide.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+The **`importmap`** value of the [`type`](/en-US/docs/Web/HTML/Element/script/type) attribute of the [`<script>` element](/en-US/docs/Web/HTML/Element/script) indicates that the body of the element contains an import map.
+
+An import map is a JSON object that allows developers to control how the browser resolves module specifiers when importing [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules).
+It provides a mapping between text that the developer wants to use in a module specifier in the [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import) or [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import), and corresponding values that will replace the text when the specifier is resolved into an absolute URL.
```
```suggestion
It provides a mapping between the text used as the module specifier in an [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import) or [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import), and the corresponding value that will replace the text when resolving the specifier.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+Note that the import map applies only to module specifiers in the [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import) or [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import); it does not apply to the path specified in the `src` attribute of a `<script>` element.
+
+For more information see the guide: [JavaScript modules > Importing modules using import maps](/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps).
```
```suggestion
For more information, see the [Importing modules using import maps](/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps) section in the JavaScript modules guide.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+The `src`, `async`, `nomodule`, `defer`, `crossorigin`, `integrity`, and `referrerpolicy` attributes must not be specified.
+
+Only the first import map in the document with an inline definition is processed; any additional import maps and external import maps are ignored.
```
FWIW, the "external import maps" looks slightly redundant: they are just `<script>` with `src`, which is described in the previous point.

hamishwillee replied:
It is, but I'm OK with this. I wanted to make it clear what would happen if you did any of the specific "bad things" so they are in this line together.

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+- Top level keys other than `imports` or `scopes`.
+- Module specifier maps where a key has a trailing backslash, but the corresponding value does not.
+- Empty module specifier keys.
+- Addresses that are not strings, or which cannot be resolved to an address.
+- Addresses that are prefixes that cannot be resolved to an absolute address.
```
Are the "addresses" here referring to the key or value? The idea of "address" also doesn't sound familiar in JS modules.

hamishwillee replied:
The value. Essentially it is saying the value must be a string, and that you have to resolve be able to resolve it to a URL for a module or a valid URL prefix (i.e. most of a URL, but not necessarily the final module name).

```suggestion
- Module specifier values that are not strings, or that cannot be resolved to an absolute URL.
- Module specifier values that are prefixes that cannot be resolved to a URL.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+## Description
+
+When importing a [JavaScript module](/en-US/docs/Web/JavaScript/Guide/Modules), both the [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import) and [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import) have a "module specifier" that indicates the module to be imported.
+A browser must be able to resolve this specifier to an absolute URL in order to be able to import the module.
```
```suggestion
A browser must be able to resolve this specifier to an absolute URL in order to import the module.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
````diff
@@ -0,0 +1,210 @@
+import { name as circleName } from "https://example.com/shapes/circle.js";
+```
+
+Import maps allow developers to specify (almost) any text they want in the module specifier; the map provides a corresponding value that will replace the text when the address of the module is resolved.
````
```suggestion
Import maps allow developers to specify (almost) any text they want in the module specifier; the map provides a corresponding value that will replace the text when the module specifier is resolved.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+{
+  "imports": {
+    "square": "./module/shapes/square.js",
+    "circle": "https://example.com/shapes/circle.js",
```
Didn't try but can you use trailing commas?

hamishwillee replied:
Probably not - I think JSON requires they not exist, and the spec for this feature says "valid json".

```suggestion
    "circle": "https://example.com/shapes/circle.js"
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+### Paths in the module specifier map key
+
+The module specifier does not have to be a bare module.
```
Would be good to define "bare module". The definition would be tricky to get right, though.

hamishwillee replied:
Yes, how about this: https://github.com/mdn/content/pull/22300/files#r1046446905 (see below)

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+If there are several module specifier keys in a module specifier map that might match, then the most specific key will be selected (i.e. the one with the longer path/value).
+
+Note that a module specifier of `./foo/../js/app.js` would be resolved to `./js/app.js` before matching.
```
```suggestion
A module specifier of `./foo/../js/app.js` would be resolved to `./js/app.js` before matching.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+The browser will fall back to checking the module specifier map in the `imports` key if a matching module specifier is not found in any of the matching scoped paths.
+
+Note that if multiple scopes match the referrer URL, then the most specific scope path is used (the scope key name with the longest name).
```
```suggestion
If multiple scopes match the referrer URL, then the most specific scope path is used (the scope key name with the longest name).
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+The browser will fall back to checking the module specifier map in the `imports` key if a matching module specifier is not found in any of the matching scoped paths.
+
+Note that if multiple scopes match the referrer URL, then the most specific scope path is used (the scope key name with the longest name).
+The browsers will fallback to the next most specific scoped path if there is no matching specifier, and so on, with a final fallback to the module specifier map in the `imports` key.
```
```suggestion
The browser falls back to the next most specific scoped path if there is no matching specifier, and so on, eventually falling back to the module specifier map in the `imports` key.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
````diff
@@ -0,0 +1,210 @@
+</script>
+```
+
+The browser will fall back to checking the module specifier map in the `imports` key if a matching module specifier is not found in any of the matching scoped paths.
````
This seems redundant given the following paragraph?

hamishwillee replied:
Good point.

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+The following is a "formal" definition of the import map JSON representation.
+
+The import map must be a valid JSON object that can define at most two optional keys: `imports` and `scopes`, and which may be empty.
```
```suggestion
The import map must be a valid JSON object that can define at most two optional keys: `imports` and `scopes`. Each key's value must be an object, which may be empty.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+- `imports` {{optional_inline}}
+
+  - : The value is a "module specifier map", which provides the mappings between module specifier text that might appear in an `import` statement or `import()` operator, and the text that will replace it when the modifier is resolved.
```
```suggestion
  - : The value is a "module specifier map", which provides the mappings between module specifier text that might appear in an `import` statement or `import()` operator, and the text that will replace it when the specifier is resolved.
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+    This is the fallback map that is searched for matching module specifiers if no `scopes` path URLs match, or if module specifier maps in matching `scopes` paths do not contain a key that matches the module specifier.
+
+    - `<module specifier map>`
```
This seems like format syntax notation? Maybe we can get rid of this DL since it just seems to define the "module specifier map" term given above.

hamishwillee replied:
I definitely want this. I'll link down to it instead.

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+Browsers generate console warnings for cases where invalid JSON is processed:
+
+- Top level keys other than `imports` or `scopes`.
+- Module specifier maps where a key has a trailing backslash, but the corresponding value does not.
+- Empty module specifier keys.
+- Addresses that are not strings, or which cannot be resolved to an address.
+- Addresses that are prefixes that cannot be resolved to an absolute address.
```
These points are all brought up again in the "module specifier map" definition, just in the negated form. Not sure if that's necessary, or if we should just say `where JSON that does not conform to the [module specifier map]() schema is processed`

hamishwillee replied:
Correct. I wasn't sure because 
1. some thing raise exceptions (invalid JSON) while others are warnings - invalid keys, values.
2. I wasn't sure the division would be so clear
3. I wanted to have a section near exceptions. I've reconsidered.

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+    - `<module specifier map>`
+      - : A "module specifier map" is a valid JSON object where the _keys_ are text that may be present in the module specifier when importing a module, and the corresponding _values_ are the URLs or paths that will replace this text when the module specifier is resolved to an address.
+
+        The modifier specifier map JSON object has the following properties:
```
```suggestion
        The modifier specifier map JSON object has the following requirements:
```

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+        - All of the values must be strings, defining either a valid absolute URL or a valid URL string that starts with `/`, `./`, or `../`.
+        - If a key ends with `/`, then the corresponding value must also end with `/`.
+          A key with a trailing `/` can be used as a prefix for when mapping (or remapping) modules addresses.
+        - If multiple keys can match the module specifier, the most specific key is used (in other words a specifier "olive/branch/" would match before "olive/").
```
```suggestion
        - The object properties' ordering is irrelevant: if multiple keys can match the module specifier, the most specific key is used (in other words, a specifier "olive/branch/" would match before "olive/").
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -119,9 +119,11 @@ Once you've exported some features out of your module, you need to import them i
 
-However, we've written the path a bit differently — we are using the dot (`.`) syntax to mean "the current location", followed by the path beyond that to the file we are trying to find. This is much better than writing out the entire relative path each time, as it is shorter, and it makes the URL portable — the example will still work if you move it to a different location in the site hierarchy.
+The _module specifier_ provides a string that the JavaScript environment can resolve to a path to the module file.
+On a browser this could be a path relative to the site root, which for our `basic-modules` example would be `/js-examples/module-examples/basic-modules`.
+However here we are instead using the dot (`.`) syntax to mean "the current location", followed by the path beyond that to the file we are trying to find. This is much better than writing out the entire relative path each time, as it is shorter, and it makes the URL portable — the example will still work if you move it to a different location in the site hierarchy.
```
```suggestion
In a browser, this could be a path relative to the site root, which for our `basic-modules` example would be `/js-examples/module-examples/basic-modules`.
However, here we are instead using the dot (`.`) syntax to mean "the current location", followed by the relative path to the file we are trying to find. This is much better than writing out the entire absolute path each time, as relative paths are shorter and make the URL portable — the example will still work if you move it to a different location in the site hierarchy.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -137,8 +139,6 @@ becomes
 
 You can see such lines in action in [`main.js`](https://github.com/mdn/js-examples/blob/master/module-examples/basic-modules/main.js).
 
-> **Note:** In some module systems, you can omit the file extension and the leading `/`, `./`, or `../` (e.g. `'modules/square'`). This doesn't work in the browser environment, as that can lead to multiple network roundtrips.
```
Still mention it, and say you need an import map to make it work? This is referring to node modules which is frankly quite different from "just omitting the leading `./`".

hamishwillee replied:
Done!

Josh-Cena reviewed (commented):
Before going into the JS guide...

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,210 @@
+
+### Paths in the module specifier map key
+
+The module specifier does not have to be a bare module.
+It can also be a URL path, an absolute URL, or a relative URL path that starts with `/`, `./`, or `../`.
```
```suggestion
Module specifier keys do not have to be single word names ("bare names").
They can also contain or end with path separators, or be absolute URLs, or be relative URL paths that start with `/`, `./`, or `../`.
```

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
````diff
@@ -0,0 +1,210 @@
+}
+</script>
+```
+
+The browser will fall back to checking the module specifier map in the `imports` key if a matching module specifier is not found in any of the matching scoped paths.
+
````
```suggestion

```

hamishwillee committed:
Apply suggestions from code review - easy ones

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,207 @@
+Only the first import map in the document with an inline definition is processed; any additional import maps and external import maps are ignored.
+An [`error` event](/en-US/docs/Web/API/Element/error_event) is fired at script elements with `type="importmap"` that are not processed (are ignored).
+
+### Exceptions
+
+- `TypeError`
+  - : The import map definition is not a JSON object, the `importmap` key is defined but its value is not a JSON object, or the `scopes` key is defined but its value is not a JSON object.
+
+### Warnings
+
+Browsers generate console warnings for cases where invalid JSON is processed:
+
+- Top level keys other than `imports` or `scopes`.
+- Module specifier maps where a key has a trailing backslash, but the corresponding value does not.
+- Empty module specifier keys.
+- Addresses that are not strings, or which cannot be resolved to an address.
+- Addresses that are prefixes that cannot be resolved to an absolute address.
```
```suggestion
### Exceptions

- `TypeError`
  - : The import map definition is not a JSON object, the `importmap` key is defined but its value is not a JSON object, or the `scopes` key is defined but its value is not a JSON object.

Browsers generate console warnings for other cases where the import map JSON does not conform to the [import map](#import_map_json_representation) schema.
```

hamishwillee committed:
Apply suggestions from code review

hamishwillee commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,199 @@
+
+- `imports` {{optional_inline}}
+
+  - : The value is a "module specifier map", which provides the mappings between module specifier text that might appear in an `import` statement or `import()` operator, and the text that will replace it when the specifier is resolved.
```
```suggestion
  - : The value is a [module specifier map(#module_specifier_map), which provides the mappings between module specifier text that might appear in an `import` statement or `import()` operator, and the text that will replace it when the specifier is resolved.
```

hamishwillee committed:
Update files/en-us/web/html/element/script/type/importmap/index.md

wbamberg reviewed (approved):
Thank you for the updates @hamishwillee !

hamishwillee committed:
Add back note

hamishwillee committed:
Update files/en-us/web/html/element/script/type/importmap/index.md

hamishwillee commented:
Thanks guys! 

@Josh-Cena I have accepted most of your comments and feedback - you might want to check a few of the "fixes". Two open discussion definitely left https://github.com/mdn/content/pull/22300#discussion_r1046347569  and https://github.com/mdn/content/pull/22300#pullrequestreview-1214420910

Josh-Cena reviewed (dismissed):
Just preventing merge before I actually read the new modules guide content

hamishwillee committed:
Update files/en-us/web/html/element/script/index.md

Josh-Cena commented:
Sorry it's getting late. I'll read the modules guide and respond to the outstanding comments tomorrow morning 😄

hamishwillee commented:
Thanks @Josh-Cena - I'm away until Friday now, so no urgency. Thanks for all your help getting this to standard.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
 
 > **Note:** Although imported features are available in the file, they are read only views of the feature that was exported. You cannot change the variable that was imported, but you can still modify properties similar to `const`. Additionally, these features are imported as live bindings, meaning that they can change in value even if you cannot modify the binding unlike `const`.
 
+## Importing modules using import maps
```
General comment: this looks a bit duplicated from the import map doc. I know you've commented about that somewhere already, so if you think it's fine, I'm also good.

hamishwillee replied:
I'm OK with it at this point. Partially because I expect you'll be pulling all this apart again at some point soon and you might take a different view.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+</script>
+```
+
+With this import map a script can now use `module/square` and `circle` as module specifiers:
````
```suggestion
With this import map, a script can now use `module/square` and `circle` as module specifiers:
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+import { name as circleName } from "circle";
+```
+
+The browser matches the module specifier in the import with a module specifier map key in the import map, in this case `circle` and `module/square`, and replaces matched keys with their associated values.
````
```suggestion
The browser matches the module specifier in the import with a module specifier map key in the import map — in this case, `circle` and `module/square` — and replaces matched keys with their associated values.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+The browser matches the module specifier in the import with a module specifier map key in the import map, in this case `circle` and `module/square`, and replaces matched keys with their associated values.
+Note that mapped relative address values are first resolved to absolute URL addresses using the [base URL](/en-US/docs/Web/HTML/Element/base) of the document containing the import map.
+
+Import maps allow modules to be imported using bare module names, similar to how they are imported with Node.js, and can also simulate importing modules from packages, and using modules without file extensions.
```
```suggestion
Import maps allow simulating importing modules from packages as if in Node.js by using bare module names, and also allow using modules without file extensions.
```

hamishwillee replied:
Not quite - I had three cases  I wanted to cover - bare modules, packages of modules where the module has a file extensions, packages of modules without file extensions. The suggestion doesn't quite work. Let me try make it simpler.

```suggestion
Import maps allow modules to be imported using bare module names (as in Node.js), and can also simulate importing modules from packages, both with and without file extensions.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+For example, in Node.js you might use the following syntax to import the "square" module.
+
+```js
+import { name, draw, reportArea, reportPerimeter } from 'square';
````
```suggestion
import { name, draw, reportArea, reportPerimeter } from "square";
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+More generally, import maps let developers write more ergonomic import code, and make it easier to manage the different versions and dependencies of modules used by a site.
+This can reduce the effort required to use the same JavaScript libraries in both browser and server.
+
+### Importing modules as bare names
```
I suggest removing this section, because I believe the text in the intro paragraph about using bare module names is self-explanatory already.

hamishwillee replied:
I want to keep this for linking and highlighting this, the major use case. I'll make the introduction more interesting so it doesn't just cover this use case.

hamishwillee replied:
Yes. I don't think we particularly needed to draw attention to that detail. But to keep you happy I have rewored :-)

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+
+Module specifier map entries, where both the specifier key and its associated value have a trailing forward slash (`/`), can be used as a path-prefix.
+This allows the remapping of a whole set of import URLs from one location to another.
+It can also be used to emulate working with "packages and modules", such as you might see in the Node ecosystem.
```
```suggestion
It can also be used to emulate working with "packages and modules", such as what you might see in the Node ecosystem.
```

hamishwillee replied:
This isn't better.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+It can also be used to emulate working with "packages and modules", such as you might see in the Node ecosystem.
+
+> **Note:** The trailing `/` indicates that the module specifier key can be substituted as _part_ of a module specifier.
+> If this is not present the browser will only match (and substitute) the whole module specifier key.
```
```suggestion
> If this is not present, the browser will only match (and substitute) the whole module specifier key.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+
+Ecosystems like Node use package managers such as npm to manage modules and their dependencies.
+The package manager ensures that each module is separated from other modules and their dependencies.
+As a result, while a complex application might include the same module multiple times, and in several different versions, users do not need to think about this complexity.
```
```suggestion
As a result, while a complex application might include the same module multiple times with several different versions in different parts of the module graph, users do not need to think about this complexity.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+The package manager ensures that each module is separated from other modules and their dependencies.
+As a result, while a complex application might include the same module multiple times, and in several different versions, users do not need to think about this complexity.
+
+> **Note:** You can do the same kind of thing using relative paths, but that forces a particular structure on your project, and means that you can't use bare module names, and so on.
```
```suggestion
> **Note:** You can also achieve version management using relative paths, but this is subpar because, among other things, this forces a particular structure on your project, and prevents you from using bare module names.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+> **Note:** You can do the same kind of thing using relative paths, but that forces a particular structure on your project, and means that you can't use bare module names, and so on.
+
+Import maps similarly allow you to have multiple versions of dependencies in your application and refer to them using the same module specifier.
+The way this works is that the `scopes` key allows you to provide module specifier maps that will be used depending on the path of the script that is performing the import.
```
```suggestion
You implement this with the `scopes` key, which allows you to provide module specifier maps that will be used depending on the path of the module performing the import.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+```
+
+With this mapping, if a script with an URL that contains `/node_modules/dependency/` imports `coolmodule`, the version in `/node_modules/some/other/location/coolmodule/index.js` will be used.
+If `coolmodule` is imported from a script with a non-matching scope path, then the "module specifier map" in `imports` will be used instead, mapping to the version in `/node_modules/coolmodule/index.js"`. The map in `imports` would also be used as a fallback if there was no matching specifier in the scoped map.
````
```suggestion
The map in `imports` is used as a fallback if there is no matching scope in the scoped map, or the matching scopes don't contain a matching specifier. For example, if `coolmodule` is imported from a script with a non-matching scope path, then the module specifier map in `imports` will be used instead, mapping to the version in `/node_modules/coolmodule/index.js`.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+If `dependency_script` changes then we have to update the import map to reflect the changed name of the module.
+We don't have to update `main_script` that depends on it, because the `main_script` import code does not change.
+
+### Dynamic import
+
+Import maps also work with [dynamic imports](#dynamic_module_loading) (see section below).
+
+This has the same restrictions as for static imports: the import must happen before any other module importing has started, and the module specifier must be resolvable to an absolute URL address for a module.
```
> the import must happen before any other module importing has started

I don't understand this constraint, and I think that defeats a main purpose of dynamic imports. An example would help.

Also I suggest moving this to the intro paragraph, because this is really orthogonal to all other subsections here, and should be an overarching thing.

hamishwillee replied:
>> the import must happen before any other module importing has started

> I don't understand this constraint, and I think that defeats a main purpose of dynamic imports. An example would help.

From [the explainer](https://github.com/WICG/import-maps#import-map-processing)

> Because they affect all imports, any import maps must be present and successfully fetched before any module resolution is done. This means that module graph fetching is blocked on import map fetching.

This is true whether you import a module dynamically or statically, because you don't know what else the module you import is going to do/import.

For similar reasons we only have inline maps definitions - the browser doesn't know what modules it needs until it has parsed the import map. If external maps were allowed then you might fetch the wrong ones unnecessarily while waiting for the map.

Very good point on moving this. I will do this afternoon.

Josh-Cena replied:
Ah, but the explainer is about dynamically creating/loading an import map, not having an import map and then dynamically importing other modules 😄 That makes more sense. I don't think we are currently discussing the possibility of dynamically-created import maps at all.

hamishwillee replied:
Yes. So while we do need to add info up the top about import map being declared before any imports happen, perhaps we don't need this section at all? Just a comment up the top that import maps work with the same for both dynamic and static imports.

Josh-Cena replied:
It's good if we can mention that import maps should be at the top, yes. And then a comment about dynamic/static imports being the same. That's exactly what I had in mind.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+
+### Feature detection
+
+You can check support for import maps using the [`HTMLScriptElement.supports()`](http://localhost:5042/en-US/docs/Web/API/HTMLScriptElement/supports) static method (which is itself broadly supported):
+
+```js
+if (HTMLScriptElement.supports && HTMLScriptElement.supports("importmap")) {
+  console.log("Browser supports import maps.");
+}
+```
````
Similarly, move this up, maybe? (Although less strong opinions on this one, since it's more like an afterthought.) Also:

```suggestion
You can check support for import maps using the [`HTMLScriptElement.supports()`](/en-US/docs/Web/API/HTMLScriptElement/supports) static method (which is itself broadly supported):

```js
if (HTMLScriptElement.supports?.("importmap")) {
  console.log("Browser supports import maps.");
}
```
```

hamishwillee replied:
I don't want to move this up because I'm unconvinced of its value. I.e. I think it will be challenging to write code that works sensibly for the case where import maps are supported and are not supported.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+}
+```
+
+If `dependency_script` changes then we have to update the import map to reflect the changed name of the module.
+We don't have to update `main_script` that depends on it, because the `main_script` import code does not change.
````
```suggestion
If `dependency_script` changes, then its hash contained in the file name changes as well. In this case, we only need to update the import map to reflect the changed name of the module.
We don't have to update the code of any JavaScript module that depends on it, because the specifier in the import statement does not change.
```

(That's what I read from this)

hamishwillee replied:
Yes, much better.

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -152,6 +155,196 @@ reportPerimeter(square1.length, reportList);
+Note that the path used to select a scope does not affect how the address is resolved.
+The value in the mapped path does not have to have to match the scopes path, and relative paths are still resolved to the base URL of the script that contains the import map.
+
+Just as for module specifier maps, you can have many scope keys, and these may contain overlapping paths.
+If multiple scopes match the referrer URL, then the most specific scope path is checked first (the longest scope key) for a matching specifier.
```
> Just as for module specifier maps

But the order-by-specificity feature of module specifier maps is not mentioned on this page, while I think it should, because it may not be immediately apparent. In Node, entries in specifier maps are tried by source order, which breaks from the JSON spec.

Josh-Cena reviewed (commented):
The guide looks really nice and made me learn something. The size does get unwieldy, but I want to eventually move all browser-dependent stuff somewhere else.

@wbamberg @hamishwillee Just as a question for the future: where do you think it should go, given it's not really a JavaScript language feature, but a platform feature? Should it be in `Web/Guide`?

hamishwillee committed:
Update files/en-us/web/html/element/script/type/importmap/index.md

hamishwillee committed:
Apply suggestions from code review - the easier ones

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

wbamberg commented:
> The guide looks really nice and made me learn something. The size does get unwieldy, but I want to eventually move all browser-dependent stuff somewhere else.
> 
> @wbamberg @hamishwillee Just as a question for the future: where do you think it should go, given it's not really a JavaScript language feature, but a platform feature? Should it be in `Web/Guide`?

This is an interesting question. A couple of things.

1. Historically we've tended to treat "the web" as a collection of more or less independently describable things, (JS, CSS, API, HTML, HTTP, basically) and each has made its own choices about what to document and how to structure documentation. Really you can see MDN best as a collection of more or less independent sites. And then we have not been good at dealing with things that cross over (like Performance or Security): we do have stuff in Web/Guide but it is badly organized and not well maintained (Learn is an exception to this but it's also really its own world). This is really a persistent weakness in our docs and one we will probably have to address some time.

2.. For this case, I think historically we've not distinguished "JavaScript" from "JavaScript on the Web" because there didn't used to be a distinction, and it has kind of crept up on us. I think actually if we did put something like this somewhere completely new (like Web/Guide) then I'd worry it would be really hard to find. I think it might be better to accommodate this in the JS docs, while organizing things so as to make it clear that this is JS-on-the-Web and is distinct from the language reference (this might also give us an affordance to talk a bit about Node, which people ask for sometimes).

Josh-Cena commented:
A new guide page in JS also sounds fair enough to me, as long as we make it clear that it only applies to the web :) (And I'd be excited if we will be able to document Node's behavior a bit as comparison)

hamishwillee committed:
Module intro complicated

hamishwillee commented:
@Josh-Cena @hamishwillee I have rewritten the intro section ... again :-(. The fact that it wasn't easy to slot in the information about matching order and the fact the map has to appear first indicates that probably it wasn't the best structure in the first place.

The new structure shows a lot more options all at once, and then presents the following sections as "expanded information".
I kept the "bare names" section because I like having it to link to, and because even though we show it in the intro now, we don't explain it.

I don't love this, but it exhausts me.

hamishwillee committed:
Merge branch 'main' into html_script_importmap_prep

hamishwillee committed:
remove dynamic import section and move feature

wbamberg commented:
> One problem, is that yari gets the bookmark tree wrong
> 
> ![image](https://user-images.githubusercontent.com/5368500/205534170-0c589f8c-660e-4842-aff6-3b08f5779a20.png)

BTW the Yari code that does this is: https://github.com/mdn/yari/blob/cdf1f18a80874a592c1ec85f36400e218ed5bab7/build/document-utils.ts#L3-L34.

I just filed https://github.com/mdn/yari/pull/7825, which if it is merged will make the breadcrumb use the `short-title` front matter key if that key exists, so some day when we define titles and short titles for for HTML attribute values, this will just work.

hamishwillee commented:
@Josh-Cena I just dismissed your old review. Ready for you to look at the new intro. Hope is OK

Josh-Cena commented on files/en-us/web/html/element/script/type/importmap/index.md:
```diff
@@ -0,0 +1,200 @@
+When importing a [JavaScript module](/en-US/docs/Web/JavaScript/Guide/Modules), both the [`import` statement](/en-US/docs/Web/JavaScript/Reference/Statements/import) and [`import()` operator](/en-US/docs/Web/JavaScript/Reference/Operators/import) have a "module specifier" that indicates the module to be imported.
+A browser must be able to resolve this specifier to an absolute URL in order to import the module.
+
+For example, the following statements import elements from the module specifier "./modules/shapes/square.js", which is a path relative to the base URL of the document, and the module specifier "https://example.com/shapes/circle.js", which is an absolute URL.
```
```suggestion
For example, the following statements import elements from the module specifier `"./modules/shapes/square.js"`, which is a path relative to the base URL of the document, and the module specifier `"https://example.com/shapes/circle.js"`, which is an absolute URL.
```

Josh-Cena commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -153,6 +156,214 @@ reportPerimeter(square1.length, reportList);
+import { name, draw, reportArea, reportPerimeter } from "square";
+```
+
+A bare name will throw a `TypeError` on a browser that does not [support import maps](/en-US/docs/Web/HTML/Element/script/type/importmap#browser_compatibility), because the browser has no way to resolve the module name to a URL.
````
It still throws if the browser supports import maps but there's no matching specifier key. I think the crucial thing is really that you have to make sure it resolves to something.

Josh-Cena reviewed (approved):
Looks much better! No significant issues on my side anymore, and I think this is good to go. It's not a JS language feature anyway, so I'm only reviewing this as a developer who uses the feature occasionally.

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

hamishwillee commented on files/en-us/web/javascript/guide/modules/index.md:
````diff
@@ -153,6 +156,214 @@ reportPerimeter(square1.length, reportList);
+import { name, draw, reportArea, reportPerimeter } from "square";
+```
+
+A bare name will throw a `TypeError` on a browser that does not [support import maps](/en-US/docs/Web/HTML/Element/script/type/importmap#browser_compatibility), because the browser has no way to resolve the module name to a URL.
+
+Using an import map we can provide the information needed by a browser to resolve the URL.
+Below you can see a map that defines a `square` module specifier key, which in this case maps to a relative address value.
````
```suggestion
To use bare names on a browser you need an import map, which provides the information needed by the browser to resolve module specifiers to URLs (JavaScript will throw a `TypeError` if it attempts to import a module specifier that can't be resolved to a module location).

Below you can see a map that defines a `square` module specifier key, which in this case maps to a relative address value.
```

hamishwillee committed:
Update files/en-us/web/javascript/guide/modules/index.md

hamishwillee commented on files/en-us/web/javascript/guide/modules/index.md:
```diff
@@ -153,6 +156,213 @@ reportPerimeter(square1.length, reportList);
+
+### Importing modules as bare names
+
+In some JavaScript environments you can use bare names for the module specifier.
+This works because the environment can resolve module names to a standard location in the file system.
+For example, in Node.js you might use the following syntax to import the "square" module.
```
```suggestion
In some JavaScript environments, such as Node.js, you can use bare names for the module specifier.
This works because the environment can resolve module names to a standard location in the file system.
For example, you might use the following syntax to import the "square" module.
```

hamishwillee committed:
Update files/en-us/web/javascript/guide/modules/index.md

hamishwillee commented:
Thanks @Josh-Cena and @wbamberg . Really appreciate your help and your patience. This was hard (at least for me). I think the reference is solid and extensible, so even if this isn't perfect we should be able to iterate it and improve it over time.
